File: src/cli/hanamictl/resources/cluster.go
/**
 * @author      Tobias Anker <tobias.anker@kitsunemimi.moe>
 *
 * @copyright   Apache License Version 2.0
 *
 *      Copyright 2022 Tobias Anker
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 */

package hanami_resources

import (
    "fmt"
    "os"
    "hanamictl/common"
    "github.com/spf13/cobra"
    "github.com/kitsudaiki/Hanami"
)

var (
    templatePath   string
    checkpointName string
    checkpointUuid string
)

var clusterHeader = []string{
    "uuid",
    "name",
    "visibility",
    "owner_id",
    "project_id",
    "created_at",
}

var clusterSaveHeader = []string{
    "uuid",
    "name",
}

var clusterRestoreHeader = []string{
    "uuid",
}

var createClusterCmd = &cobra.Command {
    Use:   "create -t TEMPLATE_PATH NAME",
    Short: "Create a new cluster.",
    Args:  cobra.ExactArgs(1),
    Run:   func(cmd *cobra.Command, args []string) {
        token := Login()
        address := os.Getenv("HANAMI_ADDRESS")
        clusterName := args[0]
        templateContent, err := os.ReadFile(templatePath)
        if err != nil {
            fmt.Println(err)
            os.Exit(1)
        }
        content, err := hanami_sdk.CreateCluster(address, token, clusterName, string(templateContent), hanamictl_common.DisableTlsVerification)
        if err != nil {
            fmt.Println(err)
            os.Exit(1)
        }
        hanamictl_common.ParseSingle(content, clusterHeader)
    },
}

var getClusterCmd = &cobra.Command {
    Use:   "get CLUSTER_UUID",
    Short: "Get information of a specific cluster.",
    Args:  cobra.ExactArgs(1),
    Run:   func(cmd *cobra.Command, args []string) {
        token := Login()
        address := os.Getenv("HANAMI_ADDRESS")
        clusterUuid := args[0]
        content, err := hanami_sdk.GetCluster(address, token, clusterUuid, hanamictl_common.DisableTlsVerification)
        if err != nil {
            fmt.Println(err)
            os.Exit(1)
        }
        hanamictl_common.ParseSingle(content, clusterHeader)
    },
}

var listClusterCmd = &cobra.Command {
    Use:   "list",
    Short: "List all cluster.",
    Run:   func(cmd *cobra.Command, args []string) {
        token := Login()
        address := os.Getenv("HANAMI_ADDRESS")
        content, err := hanami_sdk.ListCluster(address, token, hanamictl_common.DisableTlsVerification)
        if err != nil {
            fmt.Println(err)
            os.Exit(1)
        }
        hanamictl_common.ParseList(content, clusterHeader)
    },
}

var deleteClusterCmd = &cobra.Command {
    Use:   "delete CLUSTER_UUID",
    Short: "Delete a specific cluster from the backend.",
    Args:  cobra.ExactArgs(1),
    Run:   func(cmd *cobra.Command, args []string) {
        token := Login()
        address := os.Getenv("HANAMI_ADDRESS")
        clusterUuid := args[0]
        _, err := hanami_sdk.DeleteCluster(address, token, clusterUuid, hanamictl_common.DisableTlsVerification)
        if err != nil {
            fmt.Println(err)
            os.Exit(1)
        }
        fmt.Printf("successfully deleted cluster '%v'\n", clusterUuid)
    },
}

var saveClusterCmd = &cobra.Command {
    Use:   "save -n CHECKPOINT_NAME CLUSTER_UUID",
    Short: "Save cluster as checkpoint.",
    Args:  cobra.ExactArgs(1),
    Run:   func(cmd *cobra.Command, args []string) {
        token := Login()
        address := os.Getenv("HANAMI_ADDRESS")
        clusterUuid := args[0]
        content, err := hanami_sdk.SaveCluster(address, token, clusterUuid, checkpointName, hanamictl_common.DisableTlsVerification)
        if err != nil {
            fmt.Println(err)
            os.Exit(1)
        }
        hanamictl_common.ParseSingle(content, clusterSaveHeader)
    },
}

var restoreClusterCmd = &cobra.Command {
    Use:   "restore -c CHECKPOINT_UUID CLUSTER_UUID",
    Short: "Restore cluster from checkpoint.",
    Args:  cobra.ExactArgs(1),
    Run:   func(cmd *cobra.Command, args []string) {
        token := Login()
        address := os.Getenv("HANAMI_ADDRESS")
        clusterUuid := args[0]
        content, err := hanami_sdk.RestoreCluster(address, token, clusterUuid, checkpointUuid, hanamictl_common.DisableTlsVerification)
        if err != nil {
            fmt.Println(err)
            os.Exit(1)
        }
        hanamictl_common.ParseSingle(content, clusterRestoreHeader)
    },
}


var clusterCmd = &cobra.Command {
    Use:   "cluster",
    Short: "Manage cluster.",
}


func Init_Cluster_Commands(rootCmd *cobra.Command) {
    rootCmd.AddCommand(clusterCmd)

    clusterCmd.AddCommand(createClusterCmd)
    createClusterCmd.Flags().StringVarP(&templatePath, "template", "t", "", "Cluster Template (mandatory)")
    createClusterCmd.MarkFlagRequired("template")

    clusterCmd.AddCommand(getClusterCmd)

    clusterCmd.AddCommand(listClusterCmd)

    clusterCmd.AddCommand(deleteClusterCmd)

    clusterCmd.AddCommand(saveClusterCmd)
    saveClusterCmd.Flags().StringVarP(&checkpointName, "name", "n", "", "Checkpoint name (mandatory)")
    saveClusterCmd.MarkFlagRequired("name")

    clusterCmd.AddCommand(restoreClusterCmd)
    restoreClusterCmd.Flags().StringVarP(&checkpointUuid, "checkpoint", "c", "", "Checkpoint UUID (mandatory)")
    restoreClusterCmd.MarkFlagRequired("checkpoint")
}

File: src/hanami/src/core/cluster/cluster.cpp
/**
 * @file        cluster.cpp
 *
 * @author      Tobias Anker <tobias.anker@kitsunemimi.moe>
 *
 * @copyright   Apache License Version 2.0
 *
 *      Copyright 2022 Tobias Anker
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 */

#include "cluster.h"

#include <api/websocket/cluster_io.h>
#include <core/cluster/cluster_init.h>
#include <core/cluster/statemachine_init.h>
#include <core/cluster/states/task_handle_state.h>
#include <core/processing/cuda/cuda_functions.h>
#include <core/processing/logical_host.h>
#include <hanami_common/logger.h>
#include <hanami_common/statemachine.h>
#include <hanami_common/threading/thread.h>
#include <hanami_root.h>

/**
 * @brief constructor
 */
Cluster::Cluster()
{
    stateMachine = new Hanami::Statemachine();
    taskHandleState = new TaskHandle_State(this);

    m_counter.store(0, std::memory_order_relaxed);

    initStatemachine(*stateMachine, this, taskHandleState);
}

/**
 * @brief constructor to create cluster from a checkpoint
 *
 * @param data pointer to data with checkpoint
 * @param dataSize size of checkpoint in number of bytes
 */
Cluster::Cluster(const void* data, const uint64_t dataSize)
{
    m_counter.store(0, std::memory_order_relaxed);
}

/**
 * @brief destructor
 */
Cluster::~Cluster()
{
    for (Hexagon& hexagon : hexagons) {
        hexagon.attachedHost->removeHexagon(&hexagon);
    }
    delete stateMachine;
}

/**
 * @brief Cluster::incrementAndCompare
 * @param referenceValue
 */
bool
Cluster::incrementAndCompare(const uint32_t referenceValue)
{
    const int incrementedValue = m_counter.fetch_add(1, std::memory_order_relaxed);
    if (incrementedValue == referenceValue - 1) {
        m_counter.store(0, std::memory_order_relaxed);
        return true;
    }

    return false;
}

/**
 * @brief get uuid of the cluster
 *
 * @return uuid of the cluster
 */
const std::string
Cluster::getUuid()
{
    return clusterHeader.uuid.toString();
}

/**
 * @brief init new cluster
 *
 * @param clusterTemplate meta-data read from a cluster-template
 * @param uuid uuid of the cluster
 * @param host initial host to attach the hexagons. if nullptr, use the first cpu-host (default:
 * nullptr)
 *
 * @return true, if successful, else false
 */
bool
Cluster::init(const Hanami::ClusterMeta& clusterTemplate,
              const std::string& uuid,
              LogicalHost* host)
{
    return initNewCluster(this, clusterTemplate, uuid, host);
}

/**
 * @brief start a new forward train-cycle
 */
void
Cluster::startForwardCycle(const bool runNormalMode)
{
    Hanami::WorkerTask task;
    task.cluster = this;
    task.hexagonId = 0;
    task.blockId = UNINIT_STATE_16;
    task.mode = ClusterProcessingMode::TRAIN_FORWARD_MODE;
    if (runNormalMode) {
        task.mode = ClusterProcessingMode::NORMAL_MODE;
    }
    hexagons.front().attachedHost->addWorkerTaskToQueue(task);
}

/**
 * @brief start a new backward train-cycle
 */
void
Cluster::startBackwardCycle()
{
    Hanami::WorkerTask task;
    task.cluster = this;
    task.hexagonId = hexagons.size() - 1;
    task.blockId = UNINIT_STATE_16;
    task.mode = ClusterProcessingMode::TRAIN_BACKWARD_MODE;
    hexagons.back().attachedHost->addWorkerTaskToQueue(task);
}

/**
 * @brief Cluster::startReductionCycle
 */
void
Cluster::startReductionCycle()
{
    // attachedHost->addClusterToHost(this);
}

/**
 * @brief switch state of the cluster between task and direct mode
 *
 * @param newState new desired state
 *
 * @return true, if switch in statemachine was successful, else false
 */
bool
Cluster::setClusterState(const std::string& newState)
{
    if (newState == "DIRECT") {
        return goToNextState(SWITCH_TO_DIRECT_MODE);
    }

    if (newState == "TASK") {
        return goToNextState(SWITCH_TO_TASK_MODE);
    }

    return false;
}

/**
 * @brief update state of the cluster, which is caled for each finalized cluster
 */
void
Cluster::updateClusterState(const WorkerTask& task)
{
    std::lock_guard<std::mutex> guard(m_clusterStateLock);

    // TODO (kitsudaiki): check why this flag behave a bit strange
    // enableCreation = false;

    // trigger next lerning phase, if already in phase 1
    if (task.mode == ClusterProcessingMode::TRAIN_FORWARD_MODE) {
        startBackwardCycle();
    }
    else if (task.mode == ClusterProcessingMode::TRAIN_BACKWARD_MODE) {
        // reductionCounter++;
        if (reductionCounter >= 100 && clusterHeader.settings.enableReduction) {
            startReductionCycle();
            reductionCounter = 0;
        }
        else {
            sendClusterTrainEndMessage(this);
            // countSynapses(*this);
            goToNextState(NEXT);
        }
    }
    else if (task.mode == ClusterProcessingMode::REDUCTION_MODE) {
        sendClusterTrainEndMessage(this);
        goToNextState(NEXT);
    }
    else if (task.mode == ClusterProcessingMode::NORMAL_MODE) {
        sendClusterNormalEndMessage(this);
        goToNextState(NEXT);
    }
}

/**
 * @brief get actual task
 *
 * @return pointer to the actual task
 */
Task*
Cluster::getCurrentTask()
{
    std::lock_guard<std::mutex> guard(m_taskMutex);

    return m_currentTask;
}

/**
 * @brief get task-progress
 *
 * @param taskUuid UUID of the task
 *
 * @return task-progress
 */
const TaskProgress
Cluster::getProgress(const std::string& taskUuid)
{
    std::lock_guard<std::mutex> guard(m_taskMutex);

    const auto it = m_taskMap.find(taskUuid);
    if (it != m_taskMap.end()) {
        return it->second.progress;
    }

    TaskProgress progress;
    return progress;
}

/**
 * @brief get task-progress
 *
 * @param taskUuid UUID of the task
 *
 * @return task-progress
 */
const Task*
Cluster::getTask(const std::string& taskUuid)
{
    std::lock_guard<std::mutex> guard(m_taskMutex);

    const auto it = m_taskMap.find(taskUuid);
    if (it != m_taskMap.end()) {
        return &it->second;
    }

    return nullptr;
}

/**
 * @brief remove task from queue of abort the task, if actual in progress
 *
 * @param taskUuid UUID of the task
 *
 * @return always true
 */
bool
Cluster::removeTask(const std::string& taskUuid)
{
    std::lock_guard<std::mutex> guard(m_taskMutex);

    TaskState state = UNDEFINED_TASK_STATE;

    // check and update map
    auto itMap = m_taskMap.find(taskUuid);
    if (itMap != m_taskMap.end()) {
        state = itMap->second.progress.state;

        // if only queue but not activly processed at the moment, it can easily deleted
        if (state == QUEUED_TASK_STATE) {
            m_taskMap.erase(itMap);

            // update queue
            const auto itQueue = std::find(m_taskQueue.begin(), m_taskQueue.end(), taskUuid);
            if (itQueue != m_taskQueue.end()) {
                m_taskQueue.erase(itQueue);
            }

            return true;
        }

        // if task is active at the moment, then only mark it as aborted
        if (state == ACTIVE_TASK_STATE) {
            itMap->second.progress.state = ABORTED_TASK_STATE;
            return true;
        }

        // handle finished and aborted state
        if (state == FINISHED_TASK_STATE || state == ABORTED_TASK_STATE) {
            // input-data are automatically deleted, when the task was finished,
            // so removing from the list is enough
            m_taskMap.erase(itMap);
            return true;
        }
    }

    return false;
}

/**
 * @brief check if a task is finished
 *
 * @param taskUuid UUID of the task
 *
 * @return true, if task is finished, else false
 */
bool
Cluster::isFinish(const std::string& taskUuid)
{
    std::lock_guard<std::mutex> guard(m_taskMutex);

    TaskState state = UNDEFINED_TASK_STATE;

    const auto it = m_taskMap.find(taskUuid);
    if (it != m_taskMap.end()) {
        state = it->second.progress.state;
    }

    return state == FINISHED_TASK_STATE;
}

/**
 * @brief Cluster::getAllProgress
 * @param result
 */
void
Cluster::getAllProgress(std::map<std::string, TaskProgress>& result)
{
    for (const auto& [name, task] : m_taskMap) {
        result.emplace(task.uuid.toString(), task.progress);
    }
}

/**
 * @brief add new task
 *
 * @param uuid uuid of the new task for identification
 * @param task task itself
 *
 * @return false, if uuid already exist, else true
 */
Task*
Cluster::addNewTask()
{
    std::lock_guard<std::mutex> guard(m_taskMutex);

    Task newTask;
    const std::string taskUuid = newTask.uuid.toString();
    auto ret = m_taskMap.try_emplace(taskUuid, std::move(newTask));
    if (ret.second == false) {
        return nullptr;
    }

    m_taskQueue.push_back(taskUuid);

    return &m_taskMap[taskUuid];
}

/**
 * @brief run next task from the queue
 *
 * @return false, if task-queue if empty, else true
 */
bool
Cluster::getNextTask()
{
    // check number of tasks in queue
    if (m_taskQueue.size() == 0) {
        return false;
    }

    // remove task from queue
    const std::string nextUuid = m_taskQueue.front();
    m_taskQueue.pop_front();

    // init the new task
    auto it = m_taskMap.find(nextUuid);
    it->second.progress.state = ACTIVE_TASK_STATE;
    it->second.progress.startActiveTimeStamp = std::chrono::system_clock::now();
    m_currentTask = &it->second;

    return true;
}

/**
 * @brief switch statemachine of cluster to next state
 *
 * @param nextStateId id of the next state
 *
 * @return true, if statemachine switch was successful, else false
 */
bool
Cluster::goToNextState(const uint32_t nextStateId)
{
    return stateMachine->goToNextState(nextStateId);
}

/**
 * @brief finish current task
 */
TaskType
Cluster::finishTask()
{
    std::lock_guard<std::mutex> guard(m_taskMutex);

    // precheck
    if (m_currentTask != nullptr) {
        Hanami::ErrorContainer error;
        // TODO: handle error-ourpur
        m_currentTask->progress.state = FINISHED_TASK_STATE;
        m_currentTask->progress.endActiveTimeStamp = std::chrono::system_clock::now();

        m_currentTask = nullptr;
    }

    getNextTask();

    if (m_currentTask == nullptr) {
        return NO_TASK;
    }

    return m_currentTask->type;
}
